implement IPersistTimeoutsV2 interface

timbussmann opened the pull request:
Implement the IPersistTimeoutsV2 interface in order to fix loosing timeout messages when not using dtc.

see issue https://github.com/Particular/NServiceBus/issues/2885

Connects to Particular/NServiceBus#2885

timbussmann commented:
@Particular/nservicebus nhibernate experts please review

ramonsmits commented:
As discussed during the review:

* Normal operation is peek, send, delete. During peek we can do a row lock so that the delete will always succeed. Currently the peek and delete are executed in the same transaction when using connection sharing between sql transport and nhibernate persistence. Maybe this can potentially happen when running with multiple instances are fetching timeouts as these instance are competing on the timeout queue. (`ISession.Get<T>(object, LockMode.Upgrade`)
* The entity mapping is duplicate except for one property. Unknown why this is or what behavioral change this will have. Maybe refactor to one mapping due to the number of properties and using object initializer instead of a constructor which can potentially result in missing properties if the object is extended in the future.

timbussmann commented:
@ramonsmits just had a look at the `LockMode.Upgrade` when peeking but I get some unwanted behavior in the unit test. If you have two concurrent transactions creating an update lock, can they deadlock each other?

ramonsmits commented:
No, that should not be the case as yo do a SELECT FOR UPDATE, that should not result in a deadlock, it can only result in a sequential operation. 2nd transaction will have to wait before the first completes but the 1st does not have to wait for the 2nd because it already acquired a row write lock.

timbussmann commented:
@ramonsmits never mind, It seems to be an issue with SQLite, same code works on SQLExpress.

timbussmann commented:
@ramonsmits please review the latest commit, I changed the code according to your points:
* using upgrade lock for peek. This should only improve scenarios when using dtc for now, since as we saw, this version doesn't yet support connection sharing. So the only case where that lock may be held until the TryRemove is for ambient transactions.
* extracted the entity mapping so both `Peek` and `TryRemove` behave the same.

ramonsmits commented on src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_fetching_timeouts_from_storage.cs:
```diff
@@ -11,7 +11,7 @@ namespace NServiceBus.TimeoutPersisters.NHibernate.Tests
     {
         [Test]
-        public void Should_return_the_complete_list_of_timeouts()
+        public void GetNextChunk_should_return_the_complete_list_of_timeouts()
```
Why the method name change? Test is unaltered?

timbussmann replied:
to distinct between `GetNextChunk` and `Peek` tests

ramonsmits commented on src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs:
```diff
@@ -79,30 +81,97 @@ public void Should_remove_timeouts_by_id()
+        }
+
+        [Test]
+        public void RemiveTimeoutBy_should_remove_timeouts_by_sagaid()
```
Typo in method

ramonsmits commented on src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs:
```diff
@@ -79,30 +81,97 @@ public void Should_remove_timeouts_by_id()
+
+            persister.Add(timeout);
+
+            Assert.IsTrue(persister.TryRemove(timeout.Id));
```
How is this even possible? The timeout being stored has no ID assigned. Is it assigned by the database?

timbussmann replied:
yes, timeout id is always assigned by the timeout persister

ramonsmits commented on src/NServiceBus.NHibernate.Tests/TimeoutPersister/When_removing_timeouts_from_the_storage.cs:
```diff
@@ -79,30 +81,97 @@ public void Should_remove_timeouts_by_id()
+        {
+            var timeout = new TimeoutData()
+            {
+                Id = Guid.NewGuid().ToString(),
```
Conflicting behavior regarding ID assignment with test `TryRemove_should_return_false_when_timeout_already_deleted `

timbussmann replied:
oh that id assignment shouldn't be there. Thats some remainder from the SQLite tests. Going to remove that.

ramonsmits commented on src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutStorage.cs:
```diff
@@ -189,6 +180,47 @@ static string ConvertDictionaryToString(ICollection data)
             return serializer.SerializeObject(data);
         }
 
+        public TimeoutData Peek(string timeoutId)
```
As mentioned, I don't agree with this method name. Its an entity (get, delete) and not a message (peek/receive). Maybe an improvement for v6.

timbussmann replied:
happy to discuss this on the v6 solution.

timbussmann commented:
@Particular/nservicebus @ramonsmits any further concerns or can we merge this?

andreasohlund commented:
Looks good to me
